File: go/bird-watcher/bird_watcher.go
package birdwatcher
import "math"
// TotalBirdCount return the total bird count by summing
// the individual day's counts.
func TotalBirdCount(birdsPerDay []int) int {
	count := 0
    for _, v := range birdsPerDay {
        count += v
    }
	return count
}

// BirdsInWeek returns the total bird count by summing
// only the items belonging to the given week.
func BirdsInWeek(birdsPerDay []int, week int) int {
	counter := 0
    if len(birdsPerDay) < (week - 1) * 7 {
        return 0
    }
    for i:= (week - 1)* 7; i < int(math.Min(float64(len(birdsPerDay)), float64(week * 7))); i++ {
        counter += birdsPerDay[i]
    }
	return counter
}

// FixBirdCountLog returns the bird counts after correcting
// the bird counts for alternate days.
func FixBirdCountLog(birdsPerDay []int) []int {
	for i := 0; i < len(birdsPerDay); i += 2 {
        birdsPerDay[i] += 1
    }
	return birdsPerDay
}

File: java/bird-watcher/src/main/java/BirdWatcher.java

class BirdWatcher {
    private final int[] birdsPerDay;

    public BirdWatcher(int[] birdsPerDay) {
        this.birdsPerDay = birdsPerDay.clone();
    }

    public int[] getLastWeek() {
        return new int[] { 0, 2, 5, 3, 7, 8, 4};
    }

    public int getToday() {
        return this.birdsPerDay[this.birdsPerDay.length - 1];
    }

    public void incrementTodaysCount() {
        this.birdsPerDay[this.birdsPerDay.length - 1] += 1;
    }

    public boolean hasDayWithoutBirds() {
        for(int i = 0; i < this.birdsPerDay.length; i++) {
            if (this.birdsPerDay[i] == 0) {
                return true;
            }
        }
        return false;
    }

    public int getCountForFirstDays(int numberOfDays) {
        int counter = 0;
        numberOfDays = Math.min(numberOfDays, this.birdsPerDay.length);
        for (int i = 0; i < numberOfDays; i++) {
            counter += this.birdsPerDay[i];
        }
        return counter;
    }

    public int getBusyDays() {
        int busyThreshold = 5;
        int counter = 0;
        for (int i = 0; i < this.birdsPerDay.length; i++) {
            if (this.birdsPerDay[i] >= busyThreshold) {
                counter += 1;
            }
        }
        return counter;
    }
}
